### ISSUE TITLE
[Bug] Namespace=None in packages.py, prevents proper comparison between Specs.

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#1622]
@tgamblin 
The problem was that things "don't work."  See below for an (eventual) summary of how I got it working.

To simplify things, I've hacked my `openmpi/package.py` as follows:

```
#    version('2.0.0', 'cdacc800cb4ce690c1f1273cb6366674')
#    version('1.10.3', 'e2fe4513200e2aaa1500b762342c674b')
#    version('1.10.2', 'b2f43d9635d2d52826e5ef9feb97fd4c')
    version('1.10.1', 'f0fcd77ed345b7eafb431968124ba16e')
    version('1.10.0', '280cf952de68369cebaca886c5ce0304')
    version('1.8.8', '0dab8e602372da1425e9242ae37faf8c')
    version('1.6.5', '03aed2a4aa4d0b27196962a2a65fc475')
...
#    provides('mpi@:2.2', when='@1.6.5')
#    provides('mpi@:3.0', when='@1.7.5:')
    provides('mpi@:3.0', when='@1.10.1:')
#    provides('mpi@:3.1', when='@2.0.0:')
```

`spack spec mpi` gives the right answer: `openmpi@1.10.1%gcc@5.3.0~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm+verbs+vt arch=linux-SuSE11-x86_64
`

HOWEVER... now I put the following in my `packages.yaml` file:

```
    openmpi:
        paths:
            openmpi@1.10.1: /usr/local/other/SLES11.3/openmpi/1.10.1/gcc-5.3
        buildable: False
```

This causes `spack spec mpi` to crash:

```
Traceback (most recent call last):
  File "/home/rpfische/spack2/bin/spack", line 184, in <module>
    main()
  File "/home/rpfische/spack2/bin/spack", line 161, in main
    return_val = command(parser, args)
  File "/gpfsm/dnb53/rpfische/spack2/lib/spack/spack/cmd/spec.py", line 57, in spec
    spec.concretize()
  File "/gpfsm/dnb53/rpfische/spack2/lib/spack/spack/spec.py", line 1252, in concretize
    self._expand_virtual_packages(),
  File "/gpfsm/dnb53/rpfische/spack2/lib/spack/spack/spec.py", line 1169, in _expand_virtual_packages
    spec)
  File "/gpfsm/dnb53/rpfische/spack2/lib/spack/spack/concretize.py", line 129, in choose_virtual_or_external
    candidates = self._valid_virtuals_and_externals(spec)
  File "/gpfsm/dnb53/rpfische/spack2/lib/spack/spack/concretize.py", line 121, in _valid_virtuals_and_externals
    usable.sort(cmp=cmp_externals)
  File "/gpfsm/dnb53/rpfische/spack2/lib/spack/spack/concretize.py", line 108, in cmp_externals
    bi = candidates.index(b)
ValueError: list.index(x): x not in list
```

I added some debugging printout to see what's going on:

```
        def cmp_externals(a, b):
            if a.name != b.name and (not a.external or a.external_module and
                                     not b.external and b.external_module):
                # We're choosing between different providers, so
                # maintain order from provider sort
                print('candidates', candidates)
                print('a', a)
                print('b', b, type(b))
                ai = candidates.index(a)
                print('ai', ai)
                print('aa', [a == x for x in candidates])
                print('bb', [b == x for x in candidates])
                bi = candidates.index(b)
                return ai - bi
```

I also did:

```
rm -rf ~/.spack/cache
```

Here's the output:

```
candidates [openmpi@1.10.1:, mpich@1:, mpich@3:, intel-parallel-studio@cluster:+mpi, mvapich2@2.0:, mvapich2@1.9]
a mpich@1:
b openmpi@1.10.1 <class 'spack.spec.Spec'>
ai 1
aa [False, True, False, False, False, False]
bb [False, False, False, False, False, False]
```

Notice that `openmpi@1.10.1: != openmpi@1.10.1`!  Hmm, maybe there's a problem with the colon.  So I changed `package.py` to:

```
    provides('mpi@:3.0', when='@1.10.1')
```

Then deleted the cache and tried again.  Same result; now `openmpi@1.10.1 != openmpi@1.10.1`.

The key here seems to be that the `openmpi@1.10.1` read from `packages.yaml` is incompatible with the same from elsewhere.

**Comment by citibeth** [comment]
OK, I found a workaround.  Digging into the comparison code, I see that Spec._cmp_node() is used:
```
    def _cmp_node(self):
        """Comparison key for just *this node* and not its deps."""
        return (self.name,
                self.namespace,
                self.versions,
                self.variants,
                self.architecture,
                self.compiler,
                self.compiler_flags)
```
To inspect, I did the follwing:
```
                ddb = b._cmp_node()               # Spec coming from packages.yaml
                ddc = candidates[0]._cmp_node()
                print('ddb', ddb)
                print('ddc', ddc)
```
This yields:
```
ddb ('openmpi', None, [Version('1.10.1')], {}, <spack.architecture.Arch object at 0x10dfb10>, None, {})
ddc ('openmpi', 'builtin', [Version('1.10.1')], {}, <spack.architecture.Arch object at 0x10d9cd0>, None, {})
```
As you can see, the namespaces don't match between the Spec coming from `packages.yaml` (None), and the Specs coming from elsewhere ('builtin').  I have no good ideas on a real solution to this problem.  Maybe `namespace=None` should match all namespaces?

But since I don't use namespaces (for now), commenting out the following line makes things work:
```
    def _cmp_node(self):
        """Comparison key for just *this node* and not its deps."""
        return (self.name,
#                self.namespace,
                self.versions,
                self.variants,
                self.architecture,
                self.compiler,
                self.compiler_flags)
```

**Comment by citibeth** [comment]
Another workaround, I used the following for `spec_externals()` in `config.py`.  It forces the namespace of everything in `packages.yaml` to be 'builtin'.
```
def spec_externals(spec):
    """Return a list of external specs (with external directory path filled in),
       one for each known external installation."""
    allpkgs = get_config('packages')
    name = spec.name

    external_specs = []
    pkg_paths = allpkgs.get(name, {}).get('paths', None)
    pkg_modules = allpkgs.get(name, {}).get('modules', None)
    if (not pkg_paths) and (not pkg_modules):
        return []

    for external_spec, path in pkg_paths.iteritems():
        if not path:
            # skip entries without paths (avoid creating extra Specs)
            continue

        external_spec = spack.spec.Spec(external_spec, external=path)
        external_spec.namespace = 'builtin'
        if external_spec.satisfies(spec):
            external_specs.append(external_spec)

    for external_spec, module in pkg_modules.iteritems():
        if not module:
            continue

        path = get_path_from_module(module)

        external_spec = spack.spec.Spec(
            external_spec, external=path, external_module=module)
        external_spec.namespace = 'builtin'
        if external_spec.satisfies(spec):
            external_specs.append(external_spec)

    return external_specs
```

**Comment by tgamblin** [comment]
I played with this but could not reproduce it.  The line that causes the error is gone, and it looks like it was fixed by @mplegendre in #2071.  I've added another test to better cover this case in #2833.

Worth noting: the old code for `cmp_externals` shouldn't have been using `==` -- this is what @mpl fixed in #2071. It now correctly checks whether a spec `satisfies` the external spec.  It's correct that `exernalrepo.openmpi@1.10.1 != builtin.openmpi@1.10.1` and that `openmpi@1.10.1 != builtin.openmpi@1.10.1`, but both `exernalrepo.openmpi@1.10.1` and `builtin.openmpi@1.10.1` *satisfy* `openmpi@1.10.1`.  If a spec doesn't have a concrete namespace, it's treated like a wildcard in matching.  This is consistent with rules used for other attributes (variants, compilers, etc.) in Spack.
